### DISCUSSION TITLE
Debugger service target locking

### DISCUSSION THREAD

**Discussion by emad-elsaid** [go-delve/delve#3071] [Ideas]
While using the Debugger service struct. I was able to access the target using `Target()` but as it was pointed out to me [here](https://github.com/go-delve/delve/issues/3065#issuecomment-1186585944) I shouldn't be using the target without locking it.

The thing is if I do use `LockTarget()` it will block until the target is unlocked. and the mutex is not exported so I can't use `mutex.TryLock`. 

What I need to achieve is to lock the target, get some information or execute methods on it, unlock. IF the target is unlocked. and if it's now busy I want to take an alternative route, where I return cached data or show a message that the action is not allowed now until the target unlocked/halts.

Are we open to add `Debugger.TryLockTarget() bool` to the Debugger struct? it seems like a small addition I'm welling to work on if we're open for it.

**Comment by aarzilli** [comment]
I don't think we want to add more API to debugger for external use only, because we don't maintain backwards compatibility on the debugger API.

**Reply by emad-elsaid** [reply]
I see. is there a workaround to know if I'm going to be blocked using one of the methods that will lock the target or not? could `Debugger.IsRunning` be used for that?

**Reply by aarzilli** [reply]
Command has a resumeNotify channel that will be closed when the command terminates, unlocking the target.
